File: apps/api-rs/src/agent/citation_resolver.rs
use std::collections::{BTreeSet, HashMap, HashSet};

use uuid::Uuid;

use crate::models::agent::CitationOutput;
use crate::models::citation::CitationAnchor;
use crate::models::rag::{EvidencePack, RerankedChunk};

const MAX_QUOTE_CHARS: usize = 180;
const MAX_CITATIONS: usize = 6;

pub fn resolve_citations(answer: &str, evidence: &EvidencePack) -> Vec<CitationOutput> {
    let cited_indexes = cited_evidence_indexes(answer);
    if cited_indexes.is_empty() {
        return vec![];
    }
    let mut selected = Vec::new();

    for (evidence_index, chunk) in evidence.chunks.iter().enumerate() {
        let one_based = evidence_index as i32 + 1;
        if !cited_indexes.contains(&one_based) {
            continue;
        }
        selected.push((one_based, chunk));
    }

    let mut seen_docs = HashSet::new();

    selected
        .into_iter()
        .filter(|(_, chunk)| seen_docs.insert(chunk.chunk.doc_id))
        .map(|(index, chunk)| CitationOutput {
            index,
            chunk_id: chunk.chunk.chunk_id,
            doc_id: chunk.chunk.doc_id,
            doc_title: chunk.chunk.doc_title.clone(),
            page_range: chunk.chunk.page_range.clone(),
            quote: compact_quote(&chunk.chunk.content),
            score: chunk.score,
            source_status: "available".to_string(),
            anchor: Some(anchor_for_chunk(chunk)),
        })
        .take(MAX_CITATIONS)
        .collect()
}

pub fn canonicalize_citation_markers(answer: &str, evidence: &EvidencePack) -> String {
    let cited = cited_evidence_indexes(answer);
    let mut first_by_doc = HashMap::new();
    let replacements: HashMap<i32, i32> = cited
        .into_iter()
        .filter_map(|index| {
            let doc_id = evidence.chunks.get(index as usize - 1)?.chunk.doc_id;
            let canonical = *first_by_doc.entry(doc_id).or_insert(index);
            Some((index, canonical))
        })
        .collect();

    let mut result = String::with_capacity(answer.len());
    let mut rest = answer;
    while let Some(start) = rest.find('[') {
        result.push_str(&rest[..start]);
        let marker = &rest[start..];
        let Some(end) = marker.find(']') else {
            result.push_str(marker);
            return result;
        };
        let values = marker[1..end]
            .split(',')
            .map(str::trim)
            .map(str::parse::<i32>)
            .collect::<Result<Vec<_>, _>>();
        if let Ok(values) = values {
            let mut canonical = Vec::new();
            for value in values {
                let value = replacements.get(&value).copied().unwrap_or(value);
                if !canonical.contains(&value) {
                    canonical.push(value);
                }
            }
            result.push('[');
            result.push_str(
                &canonical
                    .iter()
                    .map(i32::to_string)
                    .collect::<Vec<_>>()
                    .join(","),
            );
            result.push(']');
        } else {
            result.push_str(&marker[..=end]);
        }
        rest = &marker[end + 1..];
    }
    result.push_str(rest);
    result
}

pub fn cited_evidence_indexes(answer: &str) -> BTreeSet<i32> {
    let mut indexes = BTreeSet::new();
    let mut chars = answer.chars().peekable();

    while let Some(ch) = chars.next() {
        if ch != '[' {
            continue;
        }
        let mut marker = String::new();
        while let Some(next) = chars.peek().copied() {
            chars.next();
            if next == ']' {
                break;
            }
            marker.push(next);
        }
        for part in marker.split(',') {
            if let Ok(index) = part.trim().parse::<i32>() {
                if index > 0 {
                    indexes.insert(index);
                }
            }
        }
    }

    indexes
}

fn anchor_for_chunk(chunk: &RerankedChunk) -> CitationAnchor {
    let primary = chunk.chunk.primary_anchor.as_ref();
    let page = primary
        .and_then(|a| a.page)
        .or_else(|| chunk.chunk.page_range.first().copied());
    let slide = primary
        .and_then(|a| a.slide)
        .or_else(|| metadata_i32(&chunk.chunk.metadata, "slide_start"))
        .or_else(|| metadata_i32(&chunk.chunk.metadata, "slide"))
        .or_else(|| metadata_i32(&chunk.chunk.metadata, "slide_end"));
    let kind = primary.map(|a| a.kind.clone()).unwrap_or_else(|| {
        if !chunk.chunk.table_ids.is_empty() || chunk.chunk.source_type() == "table" {
            "table_region".to_string()
        } else if slide.is_some() {
            "slide_shape".to_string()
        } else {
            "paragraph".to_string()
        }
    });
    let bbox = primary.and_then(|a| a.bbox.clone());
    let char_range = primary.and_then(|a| a.char_range.clone());
    let anchor_id = primary.map(|a| a.anchor_id);
    let parse_job_id = primary.map(|a| a.parse_job_id);

    let location_status = if bbox.is_some() || char_range.is_some() {
        "exact"
    } else if !chunk.chunk.block_ids.is_empty() || !chunk.chunk.table_ids.is_empty() {
        "structural_only"
    } else if slide.is_some() {
        "slide_only"
    } else if page.is_some() {
        "page_only"
    } else {
        "unavailable"
    };

    CitationAnchor {
        anchor_id,
        parse_job_id,
        format: primary
            .map(|a| a.format.clone())
            .unwrap_or_else(|| chunk.chunk.file_type.clone()),
        kind,
        page,
        slide,
        block_ids: chunk.chunk.block_ids.clone(),
        table_ids: chunk.chunk.table_ids.clone(),
        char_range,
        bbox,
        location_status: location_status.to_string(),
    }
}

fn metadata_i32(metadata: &serde_json::Value, key: &str) -> Option<i32> {
    metadata.get(key).and_then(|value| {
        value
            .as_i64()
            .or_else(|| value.as_str().and_then(|s| s.parse::<i64>().ok()))
            .map(|value| value as i32)
    })
}

fn compact_quote(content: &str) -> String {
    let text = strip_context_prefixes(content)
        .replace('\n', " ")
        .split_whitespace()
        .collect::<Vec<_>>()
        .join(" ");
    let mut quote = String::new();
    for ch in text.chars().take(MAX_QUOTE_CHARS) {
        quote.push(ch);
    }
    if text.chars().count() > MAX_QUOTE_CHARS {
        quote.push_str("...");
    }
    quote
}

fn strip_context_prefixes(content: &str) -> String {
    let mut lines = Vec::new();
    let mut skipped = HashSet::new();
    skipped.insert("标题路径");
    skipped.insert("页码");
    skipped.insert("Slide");

    for line in content.lines() {
        let trimmed = line.trim();
        if skipped.iter().any(|prefix| trimmed.starts_with(prefix)) {
            continue;
        }
        if trimmed.is_empty() && lines.is_empty() {
            continue;
        }
        lines.push(trimmed);
    }

    lines.join("\n").trim().to_string()
}

trait RetrievedChunkExt {
    fn source_type(&self) -> &str;
}

impl RetrievedChunkExt for crate::models::rag::RetrievedChunk {
    fn source_type(&self) -> &str {
        self.metadata
            .get("source_type")
            .and_then(|value| value.as_str())
            .unwrap_or("paragraph")
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::models::rag::RetrievedChunk;
    use crate::models::trace::RetrievalSource;

    #[test]
    fn parses_adjacent_and_grouped_citation_markers() {
        assert_eq!(
            cited_evidence_indexes("结论 [1, 2][3]"),
            BTreeSet::from([1, 2, 3])
        );
    }

    #[test]
    fn answer_without_markers_does_not_receive_automatic_citations() {
        let evidence = EvidencePack {
            chunks: vec![evidence_chunk()],
            context_text: "evidence".to_string(),
        };
        assert!(resolve_citations("没有引用标记的答案", &evidence).is_empty());
    }

    #[test]
    fn out_of_range_markers_do_not_map_to_another_chunk() {
        let evidence = EvidencePack {
            chunks: vec![evidence_chunk()],
            context_text: "evidence".to_string(),
        };
        assert!(resolve_citations("错误引用 [2]", &evidence).is_empty());
    }

    #[test]
    fn one_document_produces_one_source_and_one_marker_number() {
        let first = evidence_chunk();
        let mut second = evidence_chunk();
        second.chunk.doc_id = first.chunk.doc_id;
        second.chunk.page_range = vec![2];
        let evidence = EvidencePack {
            chunks: vec![first, second],
            context_text: "evidence".to_string(),
        };

        assert_eq!(
            resolve_citations("结论甲 [1]，结论乙 [2]", &evidence).len(),
            1
        );
        assert_eq!(
            canonicalize_citation_markers("结论甲 [1]，结论乙 [2]", &evidence),
            "结论甲 [1]，结论乙 [1]"
        );
    }

    fn evidence_chunk() -> RerankedChunk {
        RerankedChunk {
            chunk: RetrievedChunk {
                chunk_id: Uuid::new_v4(),
                doc_id: Uuid::new_v4(),
                doc_title: "测试文档".to_string(),
                file_type: "docx".to_string(),
                content: "可核验的文档事实".to_string(),
                heading_path: vec![],
                page_range: vec![1],
                block_ids: vec![],
                table_ids: vec![],
                anchor_ids: vec![],
                primary_anchor_id: None,
                anchor_quality: "page_only".to_string(),
                primary_anchor: None,
                metadata: serde_json::json!({}),
                score: 0.9,
                source: RetrievalSource::Rrf,
            },
            score: 0.9,
            rank: 1,
        }
    }
}

File: apps/api-bun/src/agent/citation_resolver.ts
// 移植自 apps/api-rs/src/agent/citation_resolver.rs
import type { CitationOutput } from '../models/agent.ts';
import type { CitationAnchor } from '../models/citation.ts';
import type { CharRange, NormalizedBBox, SourceAnchor } from '../models/source_anchor.ts';
import type { EvidencePack, RerankedChunk, RetrievedChunk } from '../models/rag.ts';
import { newUuid } from '../infra/uuid.ts';

const MAX_QUOTE_CHARS = 300;

export function resolveCitations(answer: string, evidence: EvidencePack): CitationOutput[] {
  return citationPlan(answer, evidence).citations.map(
    ({ displayIndex, chunk, sourceAnchor }) => ({
      citation_id: newUuid(),
      index: displayIndex,
      chunk_id: chunk.chunk.chunk_id,
      doc_id: chunk.chunk.doc_id,
      doc_title: chunk.chunk.doc_title,
      page_range: [...chunk.chunk.page_range],
      quote: compactQuote(sourceAnchor?.text || chunk.chunk.content),
      score: chunk.score,
      source_status: 'available',
      anchor: anchorForChunk(chunk, sourceAnchor),
    }),
  );
}

export function canonicalizeCitationMarkers(answer: string, evidence: EvidencePack): string {
  const markers = citationPlan(answer, evidence).markerDisplays;
  const rewritten = answer.replace(/\[([^\]]+)\]/g, (marker, _body: string, offset: number) => {
    const displays = markers.get(offset);
    return displays === undefined || displays.length === 0
      ? marker
      : `[${displays.join(',')}]`;
  });
  return rewritten.replace(/(\[\d+(?:,\d+)*\])(?:[ \t]*\1)+/g, '$1');
}

interface PlannedCitation {
  displayIndex: number;
  chunk: RerankedChunk;
  sourceAnchor: SourceAnchor | null;
}

interface CitationPlan {
  citations: PlannedCitation[];
  markerDisplays: Map<number, number[]>;
}

function citationPlan(answer: string, evidence: EvidencePack): CitationPlan {
  const citations: PlannedCitation[] = [];
  const markerDisplays = new Map<number, number[]>();
  const displayByLocation = new Map<string, number>();

  for (const match of answer.matchAll(/\[([^\]]+)\]/g)) {
    const offset = match.index ?? 0;
    const values = match[1]!.split(',').map((part) => parseI32Strict(part.trim()));
    if (values.some((value) => value === null)) continue;
    const displays: number[] = [];
    const claim = claimBeforeMarker(answer, offset);
    for (const evidenceIndex of values as number[]) {
      const chunk = evidence.chunks[evidenceIndex - 1];
      if (!chunk) continue;
      const sourceAnchor = selectedAnchorForChunk(chunk.chunk, claim);
      const key = citationLocationKey(chunk.chunk, sourceAnchor);
      let displayIndex = displayByLocation.get(key);
      if (displayIndex === undefined) {
        displayIndex = citations.length + 1;
        displayByLocation.set(key, displayIndex);
        citations.push({ displayIndex, chunk, sourceAnchor });
      }
      if (!displays.includes(displayIndex)) displays.push(displayIndex);
    }
    markerDisplays.set(offset, displays);
  }

  return { citations, markerDisplays };
}

function claimBeforeMarker(answer: string, offset: number): string {
  const prefix = answer.slice(Math.max(0, offset - 300), offset);
  const boundary = Math.max(
    prefix.lastIndexOf('。'),
    prefix.lastIndexOf('！'),
    prefix.lastIndexOf('？'),
    prefix.lastIndexOf('\n'),
  );
  return prefix.slice(boundary + 1).trim();
}

function citationLocationKey(chunk: RetrievedChunk, anchor: SourceAnchor | null): string {
  return `${chunk.doc_id}:${chunk.chunk_id}:${anchor?.anchor_id ?? ''}:${anchor?.page ?? ''}`;
}

export function citedEvidenceIndexes(answer: string): number[] {
  const indexes = new Set<number>();
  let cursor = 0;
  const chars = [...answer];
  while (cursor < chars.length) {
    const ch = chars[cursor];
    cursor += 1;
    if (ch !== '[') continue;
    let marker = '';
    let closed = false;
    while (cursor < chars.length) {
      const next = chars[cursor];
      cursor += 1;
      if (next === ']') {
        closed = true;
        break;
      }
      marker += next;
    }
    if (!closed) break;
    for (const part of marker.split(',')) {
      const parsed = parseI32Strict(part.trim());
      if (parsed !== null && parsed > 0) indexes.add(parsed);
    }
  }
  return [...indexes];
}

function selectedAnchorForChunk(chunk: RetrievedChunk, claim: string): SourceAnchor | null {
  const candidates = chunk.anchors.length > 0
    ? chunk.anchors
    : chunk.primary_anchor === null
      ? []
      : [chunk.primary_anchor];
  if (candidates.length === 0) return null;

  let best = chunk.primary_anchor ?? candidates[0]!;
  let bestScore = textOverlapScore(claim, best.text);
  for (const candidate of candidates) {
    const score = textOverlapScore(claim, candidate.text);
    if (score > bestScore) {
      best = candidate;
      bestScore = score;
    }
  }
  return best;
}

function textOverlapScore(left: string, right: string): number {
  const target = Array.from(left.toLowerCase().replace(/[^\p{L}\p{N}]+/gu, ''));
  const source = right.toLowerCase().replace(/[^\p{L}\p{N}]+/gu, '');
  if (target.length < 2) return source.includes(target.join('')) ? target.length : 0;
  const grams = new Set<string>();
  for (let index = 0; index < target.length - 1; index += 1) {
    grams.add(target[index]! + target[index + 1]!);
  }
  let score = 0;
  for (const gram of grams) if (source.includes(gram)) score += 1;
  return score;
}

function anchorForChunk(chunk: RerankedChunk, sourceAnchor: SourceAnchor | null): CitationAnchor {
  const page = sourceAnchor?.page
    ?? (chunk.chunk.page_range.length > 0 ? chunk.chunk.page_range[0]! : null);
  const slide = sourceAnchor?.slide
    ?? metadataI32(chunk.chunk.metadata, 'slide_start')
    ?? metadataI32(chunk.chunk.metadata, 'slide')
    ?? metadataI32(chunk.chunk.metadata, 'slide_end')
    ?? null;
  const bbox = sourceAnchor?.bbox ?? null;
  const charRange = sourceAnchor?.char_range ?? null;
  const locationStatus =
    bbox !== null || charRange !== null
      ? 'exact'
      : page !== null
        ? 'page_only'
        : slide !== null
          ? 'slide_only'
          : 'file_only';

  return {
    anchor_id: sourceAnchor?.anchor_id ?? null,
    parse_job_id: sourceAnchor?.parse_job_id ?? null,
    format: sourceAnchor?.format ?? chunk.chunk.file_type,
    kind: sourceAnchor?.kind
      ?? (chunk.chunk.table_ids.length > 0 || sourceType(chunk.chunk) === 'table'
        ? 'table_region'
        : slide !== null
          ? 'slide_shape'
          : 'paragraph'),
    page,
    slide,
    block_ids: [...chunk.chunk.block_ids],
    table_ids: [...chunk.chunk.table_ids],
    char_range: charRange,
    bbox,
    location_status: locationStatus,
  };
}

function isBBox(value: unknown): value is NormalizedBBox {
  if (typeof value !== 'object' || value === null) return false;
  const record = value as Record<string, unknown>;
  return (
    typeof record.x0 === 'number' &&
    typeof record.y0 === 'number' &&
    typeof record.x1 === 'number' &&
    typeof record.y1 === 'number'
  );
}

function isCharRange(value: unknown): value is CharRange {
  if (typeof value !== 'object' || value === null) return false;
  const record = value as Record<string, unknown>;
  return typeof record.start === 'number' && typeof record.end === 'number';
}

function metadataI32(metadata: Record<string, unknown>, key: string): number | undefined {
  const value = metadata[key];
  if (typeof value === 'number' && Number.isInteger(value)) return value;
  if (typeof value === 'string' && /^[+-]?\d+$/.test(value)) return parseI32Strict(value) ?? undefined;
  return undefined;
}

function compactQuote(content: string): string {
  const text = stripContextPrefixes(content)
    .replace(/\n/g, ' ')
    .split(/\s+/)
    .filter((part) => part.length > 0)
    .join(' ');
  const chars = [...text];
  let quote = chars.slice(0, MAX_QUOTE_CHARS).join('');
  if (chars.length > MAX_QUOTE_CHARS) quote += '...';
  return quote;
}

function stripContextPrefixes(content: string): string {
  const skipped = new Set(['标题路径', '页码', 'Slide']);
  const lines: string[] = [];
  for (const line of content.split('\n')) {
    const trimmed = line.trim();
    if ([...skipped].some((prefix) => trimmed.startsWith(prefix))) continue;
    if (trimmed.length === 0 && lines.length === 0) continue;
    lines.push(trimmed);
  }
  return lines.join('\n').trim();
}

function sourceType(chunk: RetrievedChunk): string {
  const value = chunk.metadata['source_type'];
  return typeof value === 'string' ? value : 'paragraph';
}

function parseI32Strict(text: string): number | null {
  if (!/^[+-]?\d+$/.test(text)) return null;
  const value = Number(text);
  if (value < -2147483648 || value > 2147483647) return null;
  return value;
}
